jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -181,6 +181,23 @@ ecma_is_resolver_already_called (ecma_object_t *promise_obj_p) /**< promise */
   return (ecma_promise_get_flags (promise_obj_p) & ECMA_PROMISE_ALREADY_RESOLVED) != 0;
 } /* ecma_is_resolver_already_called */
 
+/**
+ * HostPromiseRejectionTracker
+ *
+ * See also: ES11 25.6.1.9
+ */
+static void
+ecma_promise_track_rejection (ecma_object_t *obj_p, /**< promise */
+                              ecma_value_t reason) /**< reason for reject */
```

Comment:
Is this correct? According to the referenced section in the spec, the second argument of HostPromiseRejectionTracker is "operation", which can only take two values: "reject" and "handle". (Also see 25.6.1.7 (RejectPromise) and 25.6.5.4.1 (PerformPromiseThen)).

Reply:
These two values are the kind of the rejection.

"reject":
```js
Promise.reject(42)
> Uncaught (in promise) 42
```
in this case the port function is triggered with `(<new promise>, <42>)`


"handle":
```js
Promise.race()
> Uncaught (in promise) TypeError: undefined is not iterable...
```
in this case the port function is triggered with `(<new promise>, <TypeError>)`

IMO the point is the reason of the rejection not the type of the rejection. However, we can pass an enum value as a 3rd argument to indicate these two modes but i think it's superfluous.

Reply:
But I've thought that by the time control reaches the HostPromiseRejectionTracker, the promise object already "contains" the reason for rejection, so the tracker can extract it from the promise if needed.

In `ecma_promise_do_then`, step 9, that's already the case:
```c
    ecma_value_t reason = ecma_promise_get_result (promise_obj_p);
    ecma_promise_track_rejection (promise_obj_p, reason);
```

And ES11 25.6.1.7 (RejectPromise), step 3 ("Set promise.[[PromiseResult]] to reason.") also deals with this. It seems, however, that the current code (`ecma_reject_promise`) does not follow this abstract operation.

On the other hand, the "operation" argument is something that is not / cannot be deduced or extracted from either the promise object or the reason for rejection. Thus, if we want the engine to be conforming to the spec, I think we need to have that argument. (We cannot know whether a user of the engine wants to make use of it or not. So, we should err on the safe side.)

I see two options now:
1) Have the port function take two arguments, promise and operation, as per the spec. But I don't know how big changes that may implicate.
2) Have the port function take three arguments, promise, reason, and operation, and document that `ecma_promise_get_result` (`jerry_get_promise_result`?) may or may not contain the reason and therefore port implementers must not rely on that but have to use the reason argument instead.

Reply:
> And ES11 25.6.1.7 (RejectPromise), step 3 ("Set promise.[[PromiseResult]] to reason.") also deals with this. It seems, however, that the current code (ecma_reject_promise) does not follow this abstract operation.

It is set by :https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/ecma/operations/ecma-promise-object.c#L198

So probably the 2 arguments will be enough, just need to update the port implementation to obtain the reason from the rejected promise.

Reply:
Ah, OK. I've missed that.